rfc/20241021-dynamic-provider-instances.md

```diff
@@ -0,0 +1,766 @@
+			}
+			for k := range instMap {
+				instKey := addrs.StringKey(k)
+				recordInst(
+					addrs.AbsProviderInstance{
+						Module:   moduleInstAddr,
+						Provider: n.addr.Provider,
+						Alias:    n.addr.Alias,
+						Key:      instKey,
+					},
+					config,
+					EvalDataForInstanceKey(instKey, instMap),
+				)
+			}
+		default:
+			// No other situation should be possible if the config
+			// decoder is correctly implemented, so this is just a
+			// low-quality error for robustness.
+			diags = diags.Append(fmt.Errorf("provider block has invalid combination of alias and for_each arguments; it's a bug that this wasn't caught during configuration loading, so please report it"))
+		}
+	}
+
+	addRootNodeToGraph(g)
+	return g, diags.ErrWithWarnings()
+}
+```
+
+The above preserves the existing concept of initially constructing an "abstract" node type and then passing it to a "concrete" function to wrap it in a specific implementation. During all phases except the special "eval" phase used for the `tofu console` command, the final node type is [`NodeApplyableProvider](https://pkg.go.dev/github.com/opentofu/opentofu@v1.8.3/internal/tofu#NodeApplyableProvider), whose `Execute` implementation is responsible for actually starting up the provider plugin (if it isn't a builtin) and calling `ConfigureProvider` on it. The resulting `providers.Interface` value is then stored in a map keyed by `addrs.AbsProviderInstance` so that other graph nodes can find it later.
+
+The other significant change during evaluation is in the implementation of the graph nodes representing individual resource instances. Because today's OpenTofu forces a one-to-one relationship between provider configuration and provider instance, the resource instance nodes already "know" exactly which provider instance they are bound to once the initial graph has been constructed. The implementation of these node types must now perform the final step of evaluating the optional "instance key" expression in the `provider` argument, to complete the information required to construct an `addrs.AbsProviderInstance` which can be used to look up the correct `providers.Interface` to use when performing the real operations related to that resource instance.
+
+#### State and Plan Format Changes
+
+The original state snapshot format version 4 tracks a provider configuration address for each entire resource, and then all of the instances of the resource are assumed to share that same provider configuration address. To support the features in this proposal requires two changes:
+
+- Each _resource_ should record the `addrs.ConfigProviderConfig` address it was statically bound to in the most recently-applied configuration. This is technically not really a change at all, since `addrs.ConfigProviderConfig` is effectively just a rename of the current `addrs.AbsProviderConfig` and that's what OpenTofu already tracks on a per-resource basis, but it's worth noting just to complete the conceptual model.
```

Comment:
> This is technically not really a change at all

Yes, but that field has a slightly different meaning now. Previously, third parties could match resources with providers via a single field, but now resource instance providers should also be taken into account. This could cause unexpected behavior for those tools, and I would prefer them to fail instead of producing irrelevant results.

What do you think about removing this field for cases, when instance level provider should be used?

Not sure how practical this risk is though.

Reply:
Since we've explicitly put this RFC on hold in favor of updating the earlier one using static evaluation, my latest writing about this is over there in [Providers instances in the State](https://github.com/opentofu/opentofu/blob/rfc-static-eval-provider-instances-v2/rfc/20240513-static-evaluation-providers.md#providers-instances-in-the-state).

I think your same comment here applies to what I wrote in the other RFC. If you agree then please leave a similar comment over there too and I'll make sure to address it somehow before we merge those changes.

Once we've finalized the updates to the older RFC I do intend to return here and update this one to match, but right now the content in this RFC is probably not exactly aligned with the latest edits in the other RFC.